File: viewer-frontend/src/scene.ts
/**
 * Apply a moleculekit scene description to a Mol* plugin.
 *
 * The description is decided in Python (moleculekit/viewer/molstar/scene.py) so
 * that the interactive viewer and the headless renderer cannot drift: this file
 * is the single place that turns one into Mol* state, and both bundle entries
 * import it.
 */
import { MolScriptBuilder as MS } from 'molstar/lib/mol-script/language/builder'
import { StateTransforms } from 'molstar/lib/mol-plugin-state/transforms'
import { createStructureRepresentationParams } from 'molstar/lib/mol-plugin-state/helpers/structure-representation-params'
import { StructureSelectionQueries } from 'molstar/lib/mol-plugin-state/helpers/structure-selection-query'
import { StructureQueryHelper } from 'molstar/lib/mol-plugin-state/helpers/structure-query'
import { Structure, StructureSelection } from 'molstar/lib/mol-model/structure'
import { Loci } from 'molstar/lib/mol-model/loci'
import { Vec3 } from 'molstar/lib/mol-math/linear-algebra'
import { Color } from 'molstar/lib/mol-util/color'
import { ColorNames } from 'molstar/lib/mol-util/color/names'
import type { PluginContext } from 'molstar/lib/mol-plugin/context'
import type { StateBuilder, StateObjectSelector } from 'molstar/lib/mol-state'

export interface SceneSelect {
  kind: 'builtin' | 'resname' | 'atoms'
  name?: string
  names?: string[]
  indices?: number[]
}

export interface SceneComponent {
  select: SceneSelect
  representation: { type: string; size_factor?: number }
  color?: { theme?: string; uniform?: string }
  opacity?: number
}

export interface SceneLabel {
  atom: number
  position: number[]
  text: string
  size: number
  color: string
  offset: number
}

export interface Scene {
  components: SceneComponent[]
  labels?: SceneLabel[]
  tubes?: { start: number[]; end: number[]; radius: number; color: string }[]
  camera?: { direction?: number[]; up?: number[]; radius_factor?: number; focus?: SceneSelect }
  canvas?: { background?: string }
}

/** Translate an SVG colour name or hex string into a Mol* packed colour. */
function colorOf(value: string): Color {
  const named = (ColorNames as Record<string, Color>)[value.toLowerCase()]
  if (named !== undefined) return named
  return Color(parseInt(value.replace('#', ''), 16))
}

// These mirror MVS's own resolution of its static selectors, not the bare
// StructureSelectionQueries entries of the same name: MVS's static
// component type ('ligand', 'branched', ...) is resolved by
// createStructureComponent in
// mol-plugin-state/helpers/structure-component.js, which maps 'ligand' to
// StructureSelectionQueries.ligandPlusConnected (the ligand plus one layer
// of covalently/metal-connected whole residues, minus anything branched)
// and 'branched' to StructureSelectionQueries.branchedPlusConnected, not to
// the plain .ligand/.branched queries. Using the plain queries here would
// select fewer atoms than MVS does for a covalently bound ligand or a
// glycan, which is exactly the drift this shared description exists to
// remove. 'polymer', 'ion' and 'all' do resolve to their plain queries in
// that same file, so those three are used as-is.
const BUILTIN: Record<string, any> = {
  polymer: StructureSelectionQueries.polymer.expression,
  ligand: StructureSelectionQueries.ligandPlusConnected.expression,
  ion: StructureSelectionQueries.ion.expression,
  water: StructureSelectionQueries.water.expression,
  branched: StructureSelectionQueries.branchedPlusConnected.expression,
}

// MVS maps its atom_index onto Mol*'s sourceIndex, the atom's position in the
// input file, which is exactly the moleculekit atom index our description uses.
function expression(select: SceneSelect): any {
  if (select.kind === 'builtin') {
    if (select.name === 'all') return MS.struct.generator.all()
    const query = BUILTIN[select.name!]
    if (!query) throw new Error(`unknown builtin selector: ${select.name}`)
    return query
  }
  if (select.kind === 'resname') {
    return MS.struct.generator.atomGroups({
      'residue-test': MS.core.set.has([
        // MS.set(...xs) itself just collects xs into an array and hands it to
        // core.type.set; calling that directly skips spreading the array as
        // call arguments, which a large atom-index selection could overflow.
        MS.core.type.set(select.names ?? []),
        MS.ammp('label_comp_id'),
      ]),
    })
  }
  return MS.struct.generator.atomGroups({
    'atom-test': MS.core.set.has([
      MS.core.type.set(select.indices ?? []),
      MS.acp('sourceIndex'),
    ]),
  })
}

/**
 * Resolve a scene colour into `createStructureRepresentationParams` props.
 *
 * A uniform colour must land on the `color`/`colorParams` keys (the uniform
 * colour *theme*), not on `type`: `type` is the representation itself
 * (cartoon, ball-and-stick, ...), and overwriting it with a theme name there
 * makes Mol* try to use "uniform" as a representation type.
 */
function themeParams(color?: { theme?: string; uniform?: string }) {
  if (color?.uniform) {
    return { color: 'uniform', colorParams: { value: colorOf(color.uniform) } }
  }
  return { color: color?.theme ?? 'element-symbol' }
}

/**
 * Translate one MVS representation type name into Mol*'s native
 * representation type, type params and size theme.
 *
 * MVS spells representation types in its own vocabulary (snake_case names,
 * a `size_factor` field) which is not what Mol*'s own representation
 * registry uses (hyphenated names, and a `sizeFactor` type param that means
 * something else there: the sphere/cylinder radius ratio, not an overall
 * scale). Passing an MVS name straight through does not error: Mol*'s
 * `PD.Mapped` silently resolves an unrecognised name to the first
 * applicable representation for the structure, so a `ball_and_stick`
 * component drawn on a ligand-only selection quietly became an invisible
 * cartoon. This mirrors MVS's own translation
 * (`extensions/mvs/load-helpers.js`'s `representationPropsBase`) for every
 * type `build_scene()` can currently emit, so applyScene draws the same
 * picture MVS does for the same description. `size_factor` defaults to 1,
 * MVS's own default for an omitted field (`mvs-tree-representations.js`).
 */
function representationParams(representation: SceneComponent['representation']): {
  type: string
  typeParams: Record<string, unknown>
  size: string
  sizeParams: Record<string, unknown>
} {
  const value = representation.size_factor ?? 1
  switch (representation.type) {
    case 'cartoon':
      return { type: 'cartoon', typeParams: {}, size: 'uniform', sizeParams: { value } }
    case 'ball_and_stick':
      return {
        type: 'ball-and-stick',
        typeParams: { sizeFactor: 0.5, sizeAspectRatio: 0.5 },
        size: 'uniform',
        sizeParams: { value },
      }
    case 'line':
      return { type: 'line', typeParams: {}, size: 'uniform', sizeParams: { value } }
    case 'spacefill':
      return { type: 'spacefill', typeParams: {}, size: 'physical', sizeParams: { scale: value } }
    default:
      throw new Error(`applyScene cannot render representation type: ${representation.type}`)
  }
}

/** Build one component's representation and add it to `build`. */
function applyComponent(
  plugin: PluginContext,
  build: StateBuilder.Root,
  structure: StateObjectSelector,
  comp: SceneComponent
): void {
  const selection = build
    .to(structure)
    .apply(StateTransforms.Model.StructureSelectionFromExpression, {
      expression: expression(comp.select),
    })
  const repr = representationParams(comp.representation)
  const params: any = createStructureRepresentationParams(
    plugin,
    undefined,
    {
      type: repr.type as any,
      typeParams: repr.typeParams,
      size: repr.size as any,
      sizeParams: repr.sizeParams,
      ...themeParams(comp.color),
    } as any
  )
  if (comp.opacity !== undefined) {
    params.type.params.alpha = comp.opacity
  }
  selection.apply(StateTransforms.Representation.StructureRepresentation3D, params)
}

/** Label one atom, reusing the transform the interactive viewer uses for the
 * same purpose (see MolstarViewer.tsx's formal-charge labels). A label
 * attaches to a *selection*, not a bare coordinate, so this builds a
 * single-atom selection from the label's atom index first. */
function applyLabel(
  build: StateBuilder.Root,
  structure: StateObjectSelector,
  label: SceneLabel
): void {
  build
    .toRoot()
    .apply(
      StateTransforms.Model.MultiStructureSelectionFromExpression,
      {
        selections: [
          {
            key: `label-${label.atom}`,
            ref: structure.ref,
            groupId: '',
            expression: expression({ kind: 'atoms', indices: [label.atom] }),
          },
        ],
        isTransitive: true,
        label: label.text,
      },
      { dependsOn: [structure.ref] }
    )
    .apply(StateTransforms.Representation.StructureSelectionsLabel3D, {
      customText: label.text,
      textColor: colorOf(label.color),
      textSize: label.size,
      borderColor: colorOf('white'),
      borderWidth: 0.25,
      background: false,
      offsetZ: label.offset,
      scaleByRadius: false,
    } as any)
}

/** Run a selector against the realised structure and return its element loci.
 *
 * This is the same primitive `MultiStructureSelectionFromExpression` uses
 * internally (compile the expression, run it, convert the selection to a
 * loci), just invoked directly instead of through plugin state, since the
 * camera needs a bounding sphere rather than a new state object. */
function lociFor(structure: StateObjectSelector, select: SceneSelect) {
  const { selection } = StructureQueryHelper.createAndRun(
    structure.data as Structure,
    expression(select)
  )
  return StructureSelection.toLociWithSourceUnits(selection)
}

/** Move the camera onto `camera`'s focus, direction and radius. */
async function applyCamera(
  plugin: PluginContext,
  structure: StateObjectSelector,
  camera: NonNullable<Scene['camera']>
): Promise<void> {
  const canvas3d = plugin.canvas3d!
  const loci = camera.focus
    ? lociFor(structure, camera.focus)
    : Structure.toStructureElementLoci(structure.data as Structure)
  const sphere = Loci.getBoundingSphere(loci)!
  const dir = camera.direction
    ? Vec3.create(camera.direction[0], camera.direction[1], camera.direction[2])
    : undefined
  const up = camera.up ? Vec3.create(camera.up[0], camera.up[1], camera.up[2]) : undefined
  const radius = sphere.radius * (camera.radius_factor ?? 1)
  const snapshot = canvas3d.camera.getFocus(sphere.center, radius)
  // getFocus honours `dir` and `up` only up to sign: it runs them through
  // Vec3.matchDirection, which flips them into the hemisphere the camera
  // already looks along. Front and back, left and right, top and bottom then
  // all resolve to one image. Place the camera from the vectors directly.
  // `direction` points from the camera toward the target, so the position is
  // that far back along it.
  if (dir) {
    const distance = Vec3.distance(snapshot.position!, snapshot.target!)
    const offset = Vec3.setMagnitude(Vec3.zero(), dir, distance)
    snapshot.position = Vec3.sub(Vec3.zero(), snapshot.target!, offset)
  }
  if (up) snapshot.up = Vec3.clone(up)
  // Go through the pending-reset mechanism instead of camera.setState
  // directly: committing new representations (just above, in applyScene)
  // schedules Mol*'s own automatic camera-fit-to-scene reset
  // (canvas3d.js's resolveCameraReset), which runs later in the animation
  // loop and recomputes target/radius from the whole visible scene,
  // overwriting a direct setState call outright. requestCameraReset queues
  // our snapshot as `nextCameraResetSnapshot`, which resolveCameraReset
  // merges *over* its own auto-fit result instead of racing it -- the same
  // mechanism Mol*'s CameraManager.focusSphere (and MVS's own camera
  // handling) uses.
  // durationMs 0: an animated transition would let a screenshot land
  // mid-interpolation, which is how repeated renders stopped being identical
  // once before.
  canvas3d.requestCameraReset({ snapshot, durationMs: 0 })
}

/**
 * Apply a scene description to a Mol* plugin for one already-loaded structure.
 *
 * Builds every component's representation, adds formal-charge-style labels,
 * sets the canvas background and moves the camera, in that order. `tubes` is
 * not supported here: neither `render()` nor the interactive viewer ever
 * populates it, since it only exists for the MolViewSpec-only docs/notebook
 * path (see mvs.py), so a non-empty `tubes` list is rejected rather than
 * silently dropped.
 *
 * @param plugin the plugin whose state is built into
 * @param structure a Structure state selector for the molecule being described
 * @param scene the description produced by build_scene() in Python
 */
export async function applyScene(
  plugin: PluginContext,
  structure: StateObjectSelector,
  scene: Scene
): Promise<void> {
  if (scene.tubes?.length) {
    throw new Error('applyScene cannot render tubes; highlight_bonds is MVS-only')
  }

  const build = plugin.state.data.build()
  for (const comp of scene.components) {
    applyComponent(plugin, build, structure, comp)
  }
  for (const label of scene.labels ?? []) {
    applyLabel(build, structure, label)
  }
  await build.commit()

  if (scene.canvas?.background && plugin.canvas3d) {
    const hex = scene.canvas.background
    plugin.canvas3d.setProps({ renderer: { backgroundColor: colorOf(hex) } })
  }

  if (scene.camera) {
    await applyCamera(plugin, structure, scene.camera)
  }
}

File: moleculekit/viewer/molstar/scene.py
"""The rule that decides what a Mol* scene looks like, as plain data.

Both the interactive viewer and the headless renderer consume the dict this
module produces, so they cannot drift: one decides, one applies. MolViewSpec is
one possible encoding of the same description (see mvs.py) and is used by the
notebook viewer and the docs theme, which cannot reach the shared bundle.

The vocabulary is MolViewSpec's: selector names, representation type names and
Mol* colour theme names are spelled exactly as MVS spells them, so translating
in either direction is mechanical.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

import numpy as np

from moleculekit.residues import (
    NUCLEIC_RESIDUE_NAMES_WITH_VARIANTS,
    PROTEIN_RESIDUE_NAMES_WITH_VARIANTS,
)

if TYPE_CHECKING:
    from moleculekit.molecule import Molecule

logger = logging.getLogger(__name__)

# These moved here from mvs.py: scene.py owns the rule, mvs.py re-exports them.
MIN_CARTOON_RESIDUES = 6
BALL_AND_STICK_SIZE_FACTOR = 0.6
MAX_FORMAL_CHARGE_LABELS = 200

#: Fields a Labels representation can write, beyond the molecule's own per-atom
#: arrays. The names on the left are what other viewers call them.
LABEL_FIELD_ALIASES = {
    "residuename": "resname",
    "residueindex": "resid",
    "atomname": "name",
    "chainid": "chain",
}
_BALL_AND_STICK_SELECTORS = ("ligand", "ion", "water", "branched")

# Which resnames count as canonical polymer, deciding cartoon versus
# ball-and-stick. Derived from residues.py so the two cannot drift: the
# WITH_VARIANTS sets are exactly "canonical residue, including force-field
# renames" (HIS -> HID/HIE/HIP, CYS -> CYX/CYM, and so on).
STANDARD_POLYMER_RESNAMES = (
    PROTEIN_RESIDUE_NAMES_WITH_VARIANTS
    | NUCLEIC_RESIDUE_NAMES_WITH_VARIANTS
    # Spellings residues.py does not carry: ARN is the neutral-arginine name
    # some force fields use (residues.py spells it AR0), and these RNA and
    # deoxyuridine names appear in older PDB-derived files. Dropping them would
    # push those residues out of the cartoon into ball-and-stick.
    | {"ARN", "DU", "RA", "RC", "RG", "RU"}
)

DEFAULT_DIRECTION = (0.0, 0.0, -1.0)
DEFAULT_UP = (0.0, 1.0, 0.0)

# `direction` points from the camera position to the target, so "top" (looking
# down from above) is -y, which is Rx(-90) applied to DEFAULT_DIRECTION.
ORIENTATION_PRESETS = {
    "front": (0.0, 0.0, 0.0),
    "back": (0.0, 180.0, 0.0),
    "left": (0.0, -90.0, 0.0),
    "right": (0.0, 90.0, 0.0),
    "top": (-90.0, 0.0, 0.0),
    "bottom": (90.0, 0.0, 0.0),
}


def _count_standard_polymer_residues(mol) -> int:
    seen: dict = {}
    for resid, ins, chain, segid, resname in zip(
        mol.resid.tolist(),
        mol.insertion.tolist(),
        mol.chain.tolist(),
        mol.segid.tolist(),
        mol.resname.tolist(),
    ):
        seen[(resid, ins, chain, segid)] = resname
    return sum(1 for rn in seen.values() if rn in STANDARD_POLYMER_RESNAMES)


def rotation_to_direction_up(rotate):
    """Resolve a rotation into the MVS ``direction`` and ``up`` vectors.

    Parameters
    ----------
    rotate : str or tuple of float or None
        A preset name from ``ORIENTATION_PRESETS``, a tuple of ``(rx, ry, rz)``
        rotations in degrees applied about the x, y and z axes in that order, or
        None for the default view.

    Returns
    -------
    direction : tuple of float
        Unit vector from the camera position toward the target.
    up : tuple of float
        Unit vector controlling the roll about ``direction``.

    Raises
    ------
    ValueError
        If ``rotate`` is a string that names no known preset.
    """
    if rotate is None:
        return DEFAULT_DIRECTION, DEFAULT_UP

    if isinstance(rotate, str):
        key = rotate.lower()
        if key not in ORIENTATION_PRESETS:
            raise ValueError(
                f"Unknown orientation {rotate!r}. Use one of "
                f"{sorted(ORIENTATION_PRESETS)} or a (rx, ry, rz) tuple in degrees."
            )
        rotate = ORIENTATION_PRESETS[key]

    rx, ry, rz = (np.deg2rad(float(angle)) for angle in rotate)
    cx, sx = np.cos(rx), np.sin(rx)
    cy, sy = np.cos(ry), np.sin(ry)
    cz, sz = np.cos(rz), np.sin(rz)
    rot_x = np.array([[1, 0, 0], [0, cx, -sx], [0, sx, cx]])
    rot_y = np.array([[cy, 0, sy], [0, 1, 0], [-sy, 0, cy]])
    rot_z = np.array([[cz, -sz, 0], [sz, cz, 0], [0, 0, 1]])
    rot = rot_z @ rot_y @ rot_x

    direction = rot @ np.array(DEFAULT_DIRECTION)
    up = rot @ np.array(DEFAULT_UP)
    return tuple(float(v) for v in direction), tuple(float(v) for v in up)


_ELEMENT = {"theme": "element-symbol"}


def _builtin(name: str) -> dict:
    return {"kind": "builtin", "name": name}


def _atoms(indices) -> dict:
    return {"kind": "atoms", "indices": [int(i) for i in indices]}


def _ball_and_stick(select: dict, color: dict) -> dict:
    return {
        "select": select,
        "representation": {
            "type": "ball_and_stick",
            "size_factor": BALL_AND_STICK_SIZE_FACTOR,
        },
        "color": color,
    }


def _automatic_components(mol) -> list[dict]:
    """The scene shown when the user has set no representations."""
    if _count_standard_polymer_residues(mol) < MIN_CARTOON_RESIDUES:
        return [_ball_and_stick(_builtin("all"), _ELEMENT)]

    components = [
        {
            "select": _builtin("polymer"),
            "representation": {"type": "cartoon"},
            "color": {"theme": "secondary-structure"},
        }
    ]
    components += [
        _ball_and_stick(_builtin(name), _ELEMENT) for name in _BALL_AND_STICK_SELECTORS
    ]
    # Not redundant with the ligand/ion/water/branched builtins just above:
    # _bcif_bytes() (inline.py) writes only _atom_site, with no entity,
    # chem_comp or struct_conn categories. With none of those present, Mol*
    # cannot classify any atom as ligand, ion, water or branched, so all four
    # of those components draw nothing on every render; this resname-based
    # component is what actually keeps hetero atoms visible in the automatic
    # scene. Do not remove it as a believed-redundant cleanup without first
    # adding those mmCIF categories to the BinaryCIF writer and re-checking
    # coverage (see tests/test_molstar_render.py's automatic-scene
    # ligand/ion test).
    other = sorted(set(mol.resname.tolist()) - STANDARD_POLYMER_RESNAMES)
    if other:
        components.append(
            _ball_and_stick({"kind": "resname", "names": other}, _ELEMENT)
        )
    return components


def default_representations(mol) -> list[tuple]:
    """The automatic scene written as representations any viewer can draw.

    The automatic scene selects with Mol*'s builtin classifiers, which VMD and
    NGL have no equivalent of, so the split here is made with atom selection
    strings instead. It is a faithful starting point rather than a mirror: a
    residue moleculekit calls protein but that is spelled differently from the
    names in ``STANDARD_POLYMER_RESNAMES`` lands in the cartoon here and in
    ball-and-stick there.

    Parameters
    ----------
    mol : Molecule
        The molecule the scene would be built for.

    Returns
    -------
    reps : list of tuple
        ``(sel, style, color)`` triples in drawing order.
    """
    if _count_standard_polymer_residues(mol) < MIN_CARTOON_RESIDUES:
        return [("all", "CPK", "Name")]

    reps = [("protein or nucleic", "NewCartoon", "Secondary Structure")]
    hetero = "not (protein or nucleic)"
    # Skipped when it matches nothing, so a bare protein does not carry a
    # representation that build_scene would only warn about and drop.
    if mol.atomselect(hetero).any():
        reps.append((hetero, "CPK", "Name"))
    return reps


def _components_from_reps(mol, reps) -> tuple[list[dict], list[dict]]:
    """Translate user representations, which replace the automatic scene.

    Parameters
    ----------
    mol : Molecule
        The molecule the selections are resolved against.
    reps : list
        The representations to translate.

    Returns
    -------
    components : list of dict
        One scene component per drawable representation.
    labels : list of dict
        Formal charge labels contributed by ``FormalCharges`` representations.
    """
    from moleculekit.representations import Representations

    components = []
    labels = []
    dropped = []
    for rep in reps:
        translated = Representations(mol)._translateMolstar(rep)
        if translated is None:
            dropped.append(rep.sel)
            logger.warning(
                "Representation selection %r matched no atoms and was dropped.",
                rep.sel,
            )
            continue
        if translated["type"] == "label" and "label_fields" in translated:
            # Mol*'s own label representation writes a label of its own
            # choosing, so a chosen set of fields is built here instead.
            labels.extend(
                _field_labels(
                    mol,
                    translated["atom_indices"],
                    translated["label_fields"],
                    translated.get("size_factor", 1.0),
                )
            )
            continue
        if translated["type"] == "formal_charge":
            # Not a component: this draws the same per-atom "+1"/"-1" text the
            # automatic scene puts on charged atoms, restricted to the atoms
            # this representation selected.
            labels.extend(
                _labels(mol, translated["atom_indices"], translated.get("size_factor", 1.0))
            )
            continue
        color = translated.get("color")
        if color is None:
            color_spec = _ELEMENT
        elif isinstance(color, dict):
            color_spec = color
        else:
            color_spec = {"uniform": color}
        representation = {"type": translated["type"]}
        if translated["type"] == "ball_and_stick":
            # The size the automatic scene uses, so reps.addDefaults() draws
            # the same picture as setting no representations at all.
            representation["size_factor"] = BALL_AND_STICK_SIZE_FACTOR
        if "size_factor" in translated:
            representation["size_factor"] = translated["size_factor"]
        if "size_theme" in translated:
            representation["size_theme"] = translated["size_theme"]
        component = {
            "select": _atoms(translated["atom_indices"]),
            "representation": representation,
            "color": color_spec,
        }
        if "opacity" in translated:
            component["opacity"] = translated["opacity"]
        components.append(component)

    if not components and not labels:
        raise ValueError(
            "Every representation selection matched no atoms "
            f"({', '.join(repr(s) for s in dropped)}), which would render an "
            "empty scene. Check the selections, or clear mol.reps to get the "
            "automatic scene."
        )
    return components, labels


def _field_labels(mol, indices, fields, size=1.0) -> list[dict]:
    """Text beside each atom, built from the molecule's own per-atom fields.

    Mol*'s label representation writes a label of its own choosing, so anything
    else has to be built here and placed atom by atom, as the formal charge
    labels are. That costs one transform per label, which is why the same cap
    applies.

    Parameters
    ----------
    mol : Molecule
        The molecule the fields are read from.
    indices : list of int
        The atoms to label.
    fields : list of str
        Per-atom fields to write, joined by spaces.
    size : float, optional
        Scales the text.

    Returns
    -------
    labels : list of dict
        One label per atom, empty when there are more than the cap allows.

    Raises
    ------
    ValueError
        If a field is not a per-atom field of the molecule.
    """
    resolved = []
    for field in fields:
        name = LABEL_FIELD_ALIASES.get(field.lower().replace("_", ""), field.lower())
        if name == "index":
            values = np.arange(mol.numAtoms)
        else:
            values = getattr(mol, name, None)
        if values is None or len(np.atleast_1d(values)) != mol.numAtoms:
            raise ValueError(
                f"Cannot label by {field!r}: the molecule has no per-atom field "
                f"of that name. Try name, element, resname, resid, chain or index."
            )
        resolved.append(values)

    if len(indices) > MAX_FORMAL_CHARGE_LABELS:
        logger.warning(
            "Skipping labels: %d atoms exceeds the cap of %d. Each label is "
            "drawn separately, so labelling a whole structure is slow and "
            "unreadable; select the atoms worth naming.",
            len(indices),
            MAX_FORMAL_CHARGE_LABELS,
        )
        return []

    frame = mol.frame
    labels = []
    for i in indices:
        text = " ".join(str(values[i]) for values in resolved)
        labels.append(
            {
                "atom": int(i),
                "position": [float(mol.coords[i, axis, frame]) for axis in range(3)],
                "text": text,
                "size": 0.7 * size,
                "color": "black",
                "offset": 1.0,
            }
        )
    return labels


def _labels(mol, indices=None, size=1.0) -> list[dict]:
    charges = mol.formalcharge
    within = range(len(charges)) if indices is None else indices
    charged = [i for i in within if int(charges[i]) != 0]
    if not charged:
        return []
    if len(charged) > MAX_FORMAL_CHARGE_LABELS:
        logger.warning(
            "Skipping formal charge labels: %d charged atoms exceeds cap %d "
            "(likely a solvated/ionised system; show a prepared structure to "
            "keep labels meaningful).",
            len(charged),
            MAX_FORMAL_CHARGE_LABELS,
        )
        return []
    frame = mol.frame
    labels = []
    for i in charged:
        q = int(charges[i])
        labels.append(
            {
                "atom": int(i),
                "position": [float(mol.coords[i, axis, frame]) for axis in range(3)],
                "text": f"+{q}" if q > 0 else f"{q}",
                "size": 0.7 * size,
                "color": "black",
                "offset": 1.0,
            }
        )
    return labels


def _tubes(mol, highlight_bonds) -> list[dict]:
    tubes = []
    frame = mol.frame
    for sel_a, sel_b in highlight_bonds or []:
        ia = mol.atomselect(sel_a, indexes=True)
        ib = mol.atomselect(sel_b, indexes=True)
        if len(ia) != 1 or len(ib) != 1:
            raise ValueError(
                "highlight_bonds selections must each pick exactly one atom; "
                f"got {len(ia)} for {sel_a!r} and {len(ib)} for {sel_b!r}"
            )
        tubes.append(
            {
                "start": [float(v) for v in mol.coords[int(ia[0]), :, frame]],
                "end": [float(v) for v in mol.coords[int(ib[0]), :, frame]],
                "radius": 0.3,
                "color": "orange",
            }
        )
    return tubes


def focus_sphere(mols, focus_sel=None):
    """The sphere the camera should frame, across every object in the scene.

    With one molecule the browser can work this out from the selected atoms
    itself, but a selection spanning several structures cannot be expressed as
    one set of atom indices, so it is computed here from the coordinates.

    Parameters
    ----------
    mols : list
        The molecules in the scene. The frame used is each molecule's own
        ``mol.frame``.
    focus_sel : str or np.ndarray or None, optional
        Atom selection to frame. None frames everything.

    Returns
    -------
    center : list of float
        Middle of the framed atoms.
    radius : float
        Distance from the centre to the furthest framed atom, never zero.
    """
    points = []
    for mol in mols:
        coords = mol.coords[:, :, mol.frame]
        if focus_sel is not None:
            mask = mol.atomselect(focus_sel)
            if not mask.any():
                continue
            coords = coords[mask]
        points.append(coords)
    if not points:
        return None, None
    stacked = np.vstack(points)
    center = (stacked.min(axis=0) + stacked.max(axis=0)) / 2
    radius = float(np.linalg.norm(stacked - center, axis=1).max())
    # A single atom has no extent; give the camera something to frame.
    return [float(v) for v in center], max(radius, 1.0)


def build_scene(
    mol: "Molecule",
    reps=None,
    *,
    ball_and_stick_sel: "str | np.ndarray | None" = None,
    highlight_bonds: "list[tuple[str, str]] | None" = None,
    focus_sel: "str | np.ndarray | None" = None,
    rotate: "str | tuple[float, float, float] | None" = None,
    zoom: float | None = None,
    background_color: str | None = None,
    fog: float | None = None,
    clip: float | None = None,
) -> dict:
    """Describe the scene for ``mol`` as a plain dict.

    When ``reps`` is empty a cartoon is used for the polymer if ``mol`` has at
    least ``MIN_CARTOON_RESIDUES`` standard polymer residues, with ligands,
    ions, water, branched entities and non-standard residues as ball-and-stick,
    and ball-and-stick throughout otherwise. When ``reps`` is non-empty those
    representations are the whole scene, matching how the VMD and NGL backends
    treat ``mol.reps``.

    Parameters
    ----------
    mol : Molecule
        The molecule whose topology and coordinates drive the scene. The frame
        used for label and tube positions is ``mol.frame``.
    reps : list or None, optional
        Representations to render: normally the ones held in ``mol.reps``,
        together with any one-off representation added by ``view()``'s
        ``sel``, ``style`` and ``color`` arguments. An empty list or None
        gives the automatic scene.
    ball_and_stick_sel : str or np.ndarray or None, optional
        An extra atom selection to additionally draw as ball-and-stick. Ignored
        when it matches no atoms.
    highlight_bonds : list of tuple of (str, str) or None, optional
        Pairs of atom selections, each of which must pick exactly one atom. An
        orange tube is drawn between the two atoms of each pair.
    focus_sel : str or np.ndarray or None, optional
        An atom selection the camera frames on. Ignored when it matches no
        atoms.
    rotate : str or tuple of float or None, optional
        Camera orientation, as a preset name or ``(rx, ry, rz)`` in degrees.
    zoom : float or None, optional
        Camera tightness. Larger values move the camera closer.
    background_color : str or None, optional
        Canvas background as an SVG colour name or hex string.
    fog : float or None, optional
        Depth cueing strength, from 0 for none to 100 for the strongest. Fog
        fades distant geometry into the background colour. None leaves Mol*'s
        own strength.
    clip : float or None, optional
        Half-thickness in Angstrom of the slab drawn around what the camera
        frames. Geometry nearer to or further from the camera than this is cut
        away. None draws the whole structure.

    Returns
    -------
    scene : dict
        The scene description, with ``components`` always present and
        ``labels``, ``tubes``, ``camera`` and ``canvas`` present when they
        carry anything.

    Raises
    ------
    ValueError
        If every representation selection matches no atoms, if a
        ``highlight_bonds`` selection does not pick exactly one atom, if
        ``rotate`` names no known orientation preset, if ``fog`` falls outside
        0 to 100, or if ``clip`` is not positive.
    """
    if fog is not None and not 0 <= float(fog) <= 100:
        raise ValueError(f"fog must be between 0 and 100, got {fog}")
    if clip is not None and float(clip) <= 0:
        raise ValueError(f"clip must be a positive distance, got {clip}")
    if reps:
        # Labels follow the same replace-the-automatic-scene rule as the
        # components: with representations set, charges are labelled only
        # where a FormalCharges representation asks for it.
        components, labels = _components_from_reps(mol, reps)
    else:
        components = _automatic_components(mol)
        labels = _labels(mol)

    if ball_and_stick_sel is not None:
        mask = mol.atomselect(ball_and_stick_sel)
        if mask.any():
            components.append(_ball_and_stick(_atoms(mask.nonzero()[0]), _ELEMENT))

    if not components:
        # Reachable through a FormalCharges representation, which contributes
        # labels and nothing to draw. On its own it renders a blank image:
        # there is no geometry, so nothing anchors the camera either.
        raise ValueError(
            "The representations draw nothing: a FormalCharges representation "
            "labels atoms that another representation has to draw. Add one for "
            "the atoms themselves."
        )

    scene: dict = {"components": components}

    if labels:
        scene["labels"] = labels

    tubes = _tubes(mol, highlight_bonds)
    if tubes:
        scene["tubes"] = tubes

    has_camera = rotate is not None or zoom is not None
    if focus_sel is not None or has_camera:
        camera: dict = {}
        if has_camera:
            direction, up = rotation_to_direction_up(rotate)
            camera["direction"] = list(direction)
            camera["up"] = list(up)
        if zoom is not None:
            camera["radius_factor"] = 1.0 / float(zoom)
        if focus_sel is not None:
            mask = mol.atomselect(focus_sel)
            if mask.any():
                camera["focus"] = _atoms(mask.nonzero()[0])
        if camera:
            # Only alongside a camera: with none, Mol*'s own fit sets the
            # radius from the whole scene and nothing is clipped anyway.
            if clip is not None:
                camera["clip"] = float(clip)
            scene["camera"] = camera

    canvas: dict = {}
    if background_color is not None:
        canvas["background"] = background_color
    if fog is not None:
        canvas["fog"] = float(fog)
    if canvas:
        scene["canvas"] = canvas

    return scene
